File: Dart/11-Math&Geometry/Algorithms/2-SieveofEratosthenes(PrimeNumbers)/SieveofEratosthenes(PrimeNumbers).dart
// Dart program to generate prime numbers using 5 different approaches
import 'dart:math';

void main() {
  int n = 50;

  print("Prime numbers up to $n using different algorithms:");
  print("1. Basic Sieve of Eratosthenes: ${sieveOfEratosthenes(n)}");
  print("2. Optimized Sieve (Marking Multiples from Square): ${optimizedSieve(n)}");
  print("3. Segmented Sieve Algorithm: ${segmentedSieve(n)}");
  print("4. Bitwise Sieve of Eratosthenes: ${bitwiseSieve(n)}");
  print("5. Wheel Factorization Optimized Sieve: ${wheelFactorizationSieve(n)}");
}

// Approach 1: Basic Sieve of Eratosthenes
// Generates prime numbers up to n using the basic Sieve of Eratosthenes algorithm.
// It marks multiples of each prime number as non-prime.
List<int> sieveOfEratosthenes(int n) {
  List<bool> isPrime = List.filled(n + 1, true); // Initialize a boolean list to track prime numbers.
  isPrime[0] = isPrime[1] = false; // 0 and 1 are not prime.

  for (int i = 2; i * i <= n; i++) { // Iterate up to the square root of n.
    if (isPrime[i]) { // If i is prime.
      for (int j = i * i; j <= n; j += i) { // Mark multiples of i as non-prime.
        isPrime[j] = false;
      }
    }
  }
  return [for (int i = 2; i <= n; i++) if (isPrime[i]) i]; // Collect the prime numbers.
}

// Approach 2: Optimized Sieve (Marking Multiples from Square Root)
// Generates prime numbers up to n using an optimized Sieve of Eratosthenes.
// It marks multiples starting from the square of each prime, reducing redundant operations.
List<int> optimizedSieve(int n) {
  List<bool> isPrime = List.filled(n + 1, true);
  isPrime[0] = isPrime[1] = false;

  for (int i = 2; i <= sqrt(n); i++) { // Iterate up to the square root of n.
    if (isPrime[i]) {
      for (int j = i * i; j <= n; j += i) { // Mark multiples of i starting from i * i.
        isPrime[j] = false;
      }
    }
  }
  return [for (int i = 2; i <= n; i++) if (isPrime[i]) i];
}

// Approach 3: Segmented Sieve Algorithm
// Generates prime numbers up to n using the segmented Sieve of Eratosthenes.
// It divides the range into segments to reduce memory usage.
List<int> segmentedSieve(int n) {
  int limit = sqrt(n).toInt() + 1; // Calculate the limit for the initial sieve.
  List<int> primes = sieveOfEratosthenes(limit); // Generate primes up to the limit.
  List<bool> isPrime = List.filled(n + 1, true); // Initialize the main sieve.

  for (int prime in primes) { // Iterate through the primes.
    int start = max(prime * prime, (n ~/ prime) * prime); // Calculate the starting point for marking multiples.
    for (int j = start; j <= n; j += prime) { // Mark multiples of the prime.
      isPrime[j] = false;
    }
  }
  return [for (int i = 2; i <= n; i++) if (isPrime[i]) i];
}

// Approach 4: Bitwise Sieve of Eratosthenes (Efficient Memory Usage)
// Generates prime numbers using a bitwise Sieve of Eratosthenes.
// It stores prime information using bits, significantly reducing memory usage.
List<int> bitwiseSieve(int n) {
  List<int> primes = [];
  List<int> sieve = List.filled((n >> 5) + 1, 0); // Initialize a bitwise sieve.

  for (int i = 2; i <= n; i++) {
    if ((sieve[i >> 5] & (1 << (i & 31))) == 0) { // Check if i is prime.
      primes.add(i);
      for (int j = i * i; j <= n; j += i) { // Mark multiples of i as non-prime.
        sieve[j >> 5] |= (1 << (j & 31)); // Set the corresponding bit to 1.
      }
    }
  }
  return primes;
}

// Approach 5: Wheel Factorization Optimized Sieve
// Generates prime numbers using a wheel factorization optimized Sieve of Eratosthenes.
// It uses a wheel to skip multiples of small primes, improving efficiency.
List<int> wheelFactorizationSieve(int n) {
  if (n < 2) return [];
  List<bool> isPrime = List.filled(n + 1, true);
  isPrime[0] = isPrime[1] = false;
  List<int> wheel = [2, 3, 5, 7]; // Initial wheel primes.

  for (int prime in wheel) {
    if (prime * prime > n) break;
    for (int j = prime * prime; j <= n; j += prime) { // Mark multiples of the primes in the wheel.
      isPrime[j] = false;
    }
  }
  return [for (int i = 2; i <= n; i++) if (isPrime[i]) i];
}

File: Java/11-Math&Geometry/Algorithms/2-Sieve-of-Eratosthenes(PrimeNumbers)/Sieve-of-Eratosthenes(PrimeNumbers).java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class SieveOfEratosthenes {

    public static void main(String[] args) {
        int n = 50; // Find primes up to 50

        System.out.println("Sieve of Eratosthenes - Approach 1: Basic Implementation");
        List<Integer> primes1 = sieveOfEratosthenesBasic(n);
        System.out.println(primes1);

        System.out.println("\nSieve of Eratosthenes - Approach 2: Optimized with Square Root");
        List<Integer> primes2 = sieveOfEratosthenesOptimized(n);
        System.out.println(primes2);

        System.out.println("\nSieve of Eratosthenes - Approach 3: Using Boolean Array");
        List<Integer> primes3 = sieveOfEratosthenesBooleanArray(n);
        System.out.println(primes3);

        System.out.println("\nSieve of Eratosthenes - Approach 4: Bitwise Optimization");
        List<Integer> primes4 = sieveOfEratosthenesBitwise(n);
        System.out.println(primes4);

        System.out.println("\nSieve of Eratosthenes - Approach 5: Using ArrayList as Sieve");
        List<Integer> primes5 = sieveOfEratosthenesListSieve(n);
        System.out.println(primes5);
    }

    // Approach 1: Basic Implementation
    // Finds all prime numbers up to a given limit n.
    // It iterates through numbers from 2 to n, and for each number, it checks if it's prime
    // by dividing it by all smaller numbers. If a number is prime, it's added to the result list.
    // Time Complexity: O(n^2) -  because of the nested loop.
    // Space Complexity: O(n) -  to store the primes in the worst case.
    public static List<Integer> sieveOfEratosthenesBasic(int n) {
        List<Integer> primes = new ArrayList<>();
        if (n < 2) {
            return primes; // No primes less than 2
        }

        for (int i = 2; i <= n; i++) {
            boolean isPrime = true;
            for (int j = 2; j < i; j++) {
                if (i % j == 0) {
                    isPrime = false;
                    break;
                }
            }
            if (isPrime) {
                primes.add(i);
            }
        }
        return primes;
    }

    // Approach 2: Optimized with Square Root
    // An optimization of the basic approach.  Instead of checking divisibility up to i,
    // it only checks up to the square root of i.  If a number i is not prime, it can be
    // factored into two numbers, at least one of which is <= sqrt(i).
    // Time Complexity: O(n * sqrt(n)) - Improved, but still not the most efficient.
    // Space Complexity: O(n) - To store the primes.
    public static List<Integer> sieveOfEratosthenesOptimized(int n) {
        List<Integer> primes = new ArrayList<>();
        if (n < 2) {
            return primes;
        }

        for (int i = 2; i <= n; i++) {
            boolean isPrime = true;
            for (int j = 2; j <= Math.sqrt(i); j++) {
                if (i % j == 0) {
                    isPrime = false;
                    break;
                }
            }
            if (isPrime) {
                primes.add(i);
            }
        }
        return primes;
    }

    // Approach 3: Using Boolean Array (Standard Sieve of Eratosthenes)
    // The classic implementation of the Sieve of Eratosthenes.
    // It creates a boolean array, initially marking all numbers as potentially prime.
    // Then, it iterates through the array, starting from 2.  If a number is marked as prime,
    // it marks all of its multiples as not prime.
    // Time Complexity: O(n log log n) -  The most efficient time complexity for this problem.
    // Space Complexity: O(n) -  for the boolean array.
    public static List<Integer> sieveOfEratosthenesBooleanArray(int n) {
        List<Integer> primes = new ArrayList<>();
        if (n < 2) {
            return primes;
        }

        boolean[] isPrime = new boolean[n + 1];
        Arrays.fill(isPrime, true); // Assume all numbers are prime initially

        isPrime[0] = isPrime[1] = false; // 0 and 1 are not prime

        for (int p = 2; p * p <= n; p++) {
            // If p is prime, mark all multiples of p as not prime
            if (isPrime[p]) {
                for (int i = p * p; i <= n; i += p) {
                    isPrime[i] = false;
                }
            }
        }

        // Collect the prime numbers
        for (int i = 2; i <= n; i++) {
            if (isPrime[i]) {
                primes.add(i);
            }
        }
        return primes;
    }

    // Approach 4: Bitwise Optimization
    // Further optimization of the boolean array approach.  Instead of using a boolean array,
    // which uses 1 byte per element, it uses an integer array where each bit represents the
    // primality of a number.  This significantly reduces the space usage.
    // Time Complexity: O(n log log n) - Same as the standard Sieve.
    // Space Complexity: O(n/32) -  Significantly reduced due to bitwise operations.  Each integer can store 32 boolean values.
    public static List<Integer> sieveOfEratosthenesBitwise(int n) {
        List<Integer> primes = new ArrayList<>();
        if (n < 2) {
            return primes;
        }

        int[] isPrime = new int[(n / 32) + 1]; // Each int represents 32 numbers
        // (n/32) + 1  is the number of 32-bit integers needed to represent numbers up to n.

        // Function to set the bit representing the primality of a number
        // (num >> 5)  is equivalent to num / 32, determining the index of the integer in the isPrime array.
        // (1 << (num & 31)) creates a bitmask with a 1 at the position corresponding to num % 32,
        //  which is the bit position within that integer.
        //  The | operator sets that specific bit to 1, marking the number as prime.
        definePrime(isPrime, 2); // 2 is prime

        for (int p = 3; p <= n; p += 2) {
            if (isPrimeBit(isPrime, p)) {
                for (int i = p * p; i <= n; i += 2 * p) {
                    clearBit(isPrime, i); // Mark multiples of p as not prime
                }
            }
        }

        primes.add(2); // Add 2 to the result

        for (int i = 3; i <= n; i += 2) {
            if (isPrimeBit(isPrime, i)) {
                primes.add(i);
            }
        }
        return primes;
    }

    // Helper function to set a number as prime (set its corresponding bit)
    private static void definePrime(int[] isPrime, int num) {
        isPrime[num >> 5] |= (1 << (num & 31));
    }

    // Helper function to clear a number (mark as not prime)
    private static void clearBit(int[] isPrime, int num) {
        isPrime[num >> 5] &= ~(1 << (num & 31));
    }

    // Helper function to check if a number is marked as prime
    private static boolean isPrimeBit(int[] isPrime, int num) {
        return (isPrime[num >> 5] & (1 << (num & 31))) != 0;
    }

    // Approach 5: Using ArrayList as Sieve
    // This approach uses an ArrayList to represent the sieve.  The ArrayList initially contains
    // the numbers from 2 to n.  Then, it iterates through the list, removing multiples of each prime number.
    // This method can be less efficient for very large n due to the overhead of ArrayList operations
    // (removing elements).
    // Time Complexity:  O(n^2) in the worst case, but can be closer to O(n log log n) in practice,
    //                   especially with optimizations.  Removing elements from an ArrayList is O(n).
    // Space Complexity: O(n) -  for the ArrayList.
    public static List<Integer> sieveOfEratosthenesListSieve(int n) {
        List<Integer> numbers = new ArrayList<>();
        List<Integer> primes = new ArrayList<>();

        if (n < 2) return primes;

        // Initialize the list with numbers from 2 to n
        for (int i = 2; i <= n; i++) {
            numbers.add(i);
        }

        int p = 0; // Index of the current prime number
        while (p < numbers.size()) {
            int prime = numbers.get(p);
            primes.add(prime);

            // Remove multiples of the current prime.  Iterate *backwards* through the list
            // to avoid index out of bounds exceptions when removing elements.
            for (int i = numbers.size() - 1; i >= p; i--) {
                if (numbers.get(i) % prime == 0) {
                    numbers.remove(i);
                }
            }
            // p stays the same because after removing elements, the next number
            // will shift to the current index.
        }
        return primes;
    }
}

